src/route-loading.ts

```diff
@@ -0,0 +1,145 @@
+import { activationStrategy, _buildNavigationPlan } from './navigation-plan';
+import { NavigationInstruction } from './navigation-instruction';
+import { ViewPortInstruction, RouteConfig, RoutingComponent } from './interfaces';
+import { Redirect } from './navigation-commands';
+import { Router } from './router';
+import { Next } from './pipeline';
+
+export class RouteLoader {
+  loadRoute(router: Router, config: RouteConfig, navigationInstruction: NavigationInstruction): Promise<RoutingComponent> {
+    throw Error('Route loaders must implement "loadRoute(router, config, navigationInstruction)".');
+  }
+}
+
+export class LoadRouteStep {
+
+  static inject() { return [RouteLoader]; }
+  /**@internal */
+  routeLoader: RouteLoader;
+
+  constructor(routeLoader: RouteLoader) {
+    this.routeLoader = routeLoader;
+  }
+
+  run(navigationInstruction: NavigationInstruction, next: Next) {
+    return loadNewRoute(this.routeLoader, navigationInstruction)
+      .then(next)
+      .catch(next.cancel);
+  }
+}
+
+function loadNewRoute(routeLoader: RouteLoader, navigationInstruction: NavigationInstruction): Promise<any[] | void> {
+  let toLoad = determineWhatToLoad(navigationInstruction);
+  let loadPromises = toLoad.map((current) => loadRoute(
+    routeLoader,
+    current.navigationInstruction,
+    current.viewPortPlan
+  ));
+
+  return Promise.all(loadPromises);
+}
+
+interface ILoadingPlan {
```

Comment:
Maybe shift to interfaces

Reply:
The only function that uses this interface is local to this module so I thought it may as well sit in the same place